### PR TITLE
Leaves - Bri 

### PR DESCRIPTION
**Pull request by brilatimer**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | Featured is a custom method created in model Works.
Describe how you approached testing that model method. What edge cases did you come up with? | Edge case: what if there are no votes, will Spotlight be empty (yes).
What are `session` and `flash`? What is the difference between them? | Session stores a logged in user (their vote history and interaction). Flash shows various messages to the user when an issue occurs, as a warning, or as a success message.
What was one thing that you gained more clarity on through this assignment? | MVC design made more sense with this project. Helper methods continued to be useful in solving this project. 
What is the Heroku URL of your deployed application? | https://bri-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | 
The gem install minitest process had some issues that stumped me for awhile. Naming the model Media, although warned in the instructions, threw me and took a long time to fix. Suggestion: rename project to Works or Genres. Voting functionality was the most challenging to implement and a walk through via Books would have been helpful.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  root 'featured#index'
+  resources :works
+  resources :users
```
Comment (dHelmgren) [inline config/routes.rb:5]: Do you use all of these routes?

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      <%= link_to "<button>Edit</button>".html_safe, edit_work_path(work) %>
+      <%= link_to "<button>Delete</button>".html_safe, work_path(work), method: :delete, data: {confirm: "Are you sure you want to delete this media entry?"}%>
+        <% if @user && Vote.exists?(work_id:work.id,user_id:@user.id) %>
+      Voted
+    <% else  %>
+      <%= link_to "<button>Upvote</button>".html_safe, vote_by_work_id_path(work) %>
+       <% end %>
+    </ol>
+  <% end %>
+</ul>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:41]: This code repeats 3 times. It's a good candidate for a view partial

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class Work < ApplicationRecord
+  validates :creator, presence: true
+  validates :title, presence: true
+  has_many :votes
```
Comment (dHelmgren) [inline app/models/work.rb:4]: There are validations that you can put here that prevent duplicate users voting on the same work. This is the better way to handle duplicate votes.

File: app/views/works/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      <%= work[:category] %>
+      <%= work[:creator] %>
+      <%= work[:pubolcation_year] %>
+      <%= link_to "<button>Edit</button>".html_safe, edit_work_path(work) %>
+      <%= link_to "<button>Delete</button>".html_safe, work_path(work), method: :delete, data: {confirm: "Are you sure you want to delete this media entry?"}%>
+        <% if @user && Vote.exists?(work_id:work.id,user_id:@user.id) %>
+      Voted
+    <% else  %>
+      <%= link_to "<button>Upvote</button>".html_safe, vote_by_work_id_path(work) %>
+       <% end %>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:18]: I can still just punch in the URL for the `vote_by_work_id_path` and vote again.

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  def logout
+    session[:user_id] = nil
+    redirect_to root_path
+  end
+  
+  def upvote
+    @user = User.find(params[:id])
+    @user.upvote_by current_user
+    redirect_to users_path
+  end
```
Comment (dHelmgren) [inline app/controllers/users_controller.rb:59]: This breaks if we don't find a user, causing a bad page on Heroku

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | It breaks on Heroku
Can edit this work successfully, and get a flash message | no flash message
Can go to "View all media" page and see three lists of works, sorted by vote | it shows, but the index isn't sorted by vote.
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | no message
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | header only, you should make a 404 page instead.
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | You kind of hack around this, but your removal of the button doesn't keep me from casting duplicate votes.
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | no, see comment
The homepage view, all media view, and new works view use semantic HTML | It's present in your application view, but sparse elsewhere
The homepage view, all media view, and new works view use partials when appropriate | no, see comment.
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | no
Some controller, likely the `WorksController`, has a controller filter for finding a work | no
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
This submission has a lot of stuff going for it, but you also have gaps in implementation. I think a handful of important learning goals were missed here, and some important parts of the site aren't working the way that they need to be. Review my notes and let me know if you have questions.
